Adding Mish activation function

thebhatman opened the pull request:
This PR adds the Mish Activation Function and resolves https://github.com/davisking/dlib/issues/1926

davisking commented on dlib/cuda/cpu_dlib.cpp:
```diff
@@ -1467,6 +1467,48 @@ namespace dlib
+                {
+                    auto delta = 2*std::exp(d[i]) + std::exp(2*d[i]) + 2;
+                    auto omega = 4*(d[i] + 1) + 4*std::exp(2*d[i]) + std::exp(3*d[i]) + std::exp(d[i])*(4*d[i] + 6);
+                    g[i] = in[i]*std::exp(d[i])*delta/(omega*omega);
```
This kind of thing is numerically unstable, in the same way that `log(1+exp(x))` is unstable. If the inputs are not small, exp() will produce inf which will then convert to NaN in the next lines, causing training to fail.  This kind of thing needs to be written in a way that doesn't cause values of `d` with modest values (e.g. a `d` of 5000) to yield NaN.

Take a look at the graph of this function: https://www.wolframalpha.com/input/?i=derivative+of+x*tanh%28log%281%2Bexp%28x%29%29%29. Note how the gradient saturates at 1 and 0.  Again, you can use if statements similar to how `dlib::log1pexp` handles this kind of thing to make the gradient computation numerically stable.

davisking commented on dlib/cuda/cudnn_dlibapi.cpp:
```diff
@@ -162,6 +162,12 @@ namespace dlib
 
+        static cudnnActivationDescriptor_t mish_activation_descriptor()
+        {
+            thread_local cudnn_activation_descriptor des(CUDNN_ACTIVATION_MISH, CUDNN_PROPAGATE_NAN,0);
```
This isn't available in all versions of cudnn. There should be some `#if` preprocessor check around this to allow old versions of cudnn to be used.  The simplest thing to do is to provide a CUDA method that implements this stuff and call that if cudnn is too old.  You can use a grid stride loop to do this easily by basically just copying and pasting your C++ code into the cuda kernel.  For example, this is the cuda code in dlib that implements prelu:

```
        __global__ void _cuda_prelu(const float* s, float* d, size_t n, const float* pp)
        {
            const float p = *pp; 
            for (auto i : grid_stride_range(0, n))
            {
                if (s[i] > 0) 
                    d[i] = s[i];
                else
                    d[i] = p*s[i];
            }
        }

        void prelu (
            tensor& dest,
            const tensor& src, 
            const tensor& param
        )
        {
            launch_kernel(_cuda_prelu, max_jobs(dest.size()), 
                src.device(), dest.device(), src.size(), param.device());
        }
```
Nothing tricky is required. Honestly it's simpler than calling a lot of the cuDNN functions.

davisking commented on dlib/cuda/cudnn_dlibapi.h:
```diff
@@ -434,6 +434,41 @@ namespace dlib
+                    - #dest.host()[i] == src.host()[i]*std::tanh(std::log(1+std::exp(src.host()[i])))
+                - This function supports in-place operation, i.e. having
+                  is_same_object(dest, src)==true
+        !*/
```
Delete this comment.

davisking commented on dlib/cuda/cudnn_dlibapi.h:
```diff
@@ -434,6 +434,41 @@ namespace dlib
+                  assigns it to grad.
+                - This function supports in-place operation, i.e. having
+                  is_same_object(grad, gradient_input)==true
+        !*/
```
Delete this comment.  

I just noticed that these comments are here for some of these functions (this one is copied and pasted from another one obviously) and they are wrong.  The real ones are in tensor_tools.h, which have the current correct docs.  I just deleted all these from cudnn_dlibapi.h as they are duplicative with tensor_tools.h and often wrong and outdated :(

thebhatman replied:
I have removed the comments now.

davisking commented on dlib/cuda/tensor_tools.cpp:
```diff
@@ -826,6 +826,33 @@ namespace dlib { namespace tt
 
+// ----------------------------------------------------------------------------------------
+
+        void mish (
```
Fix indenting

thebhatman commented:
@davisking Thank you for an detailed review. I have added the test for Mish, and simplified the computation of Mish. I am yet to simplify the gradient computation and add the cuda kernel. I will be doing that soon. The unnecessary comments have been removed and indentation has been fixed.

thebhatman commented:
I have made the gradients zero when `x > 8` and `x < -8` . Is the current implementation numerically stable enough?

thebhatman commented:
I had initially written a kernel for mish_gradient by looking at the code for prelu_gradient. pgrad isn't required for Mish right? As it doesn't take in any parameters. So just the cuda_mish kernel should suffice right? I have removed the kernel for mish_gradient.

davisking commented:
> I had initially written a kernel for mish_gradient by looking at the code for prelu_gradient. pgrad isn't required for Mish right? As it doesn't take in any parameters. So just the cuda_mish kernel should suffice right? I have removed the kernel for mish_gradient.

Right, mish doesn't have any parameters so there is no parameter gradient.

thebhatman commented:
Is there anything more to be done @davisking ? Thanks!

davisking commented:
Sorry, I thought you were still working on it. I’ll look at this in a little bit.

davisking commented:
I just ran the unit tests (via `./dtest  --test_dnn`) with dlib built without cuda and got this error:
```
Error occurred at line 1839.
Error occurred in file /home/davis/source/dlib/dlib/test/dnn.cpp.
Failing expression was res.
Gradient error in data variable #0.  Relative error: 0.970753
expected derivative: 1.54417
output derivative:   0.045163
iteration:           0
```

davisking commented on dlib/cuda/cudnn_dlibapi.cpp:
```diff
@@ -162,6 +162,12 @@ namespace dlib
 
+        static cudnnActivationDescriptor_t mish_activation_descriptor()
+        {
+            thread_local cudnn_activation_descriptor des(CUDNN_ACTIVATION_MISH, CUDNN_PROPAGATE_NAN,0);
```
This needs to have #ifdef around it like I noted earlier.  For instance, I just tried to build this with cudnn 10.1 and it failed to build because this feature is not available.

davisking replied:
Really though, it's easier to just delete the cudnn calls and use your own cuda code.  You already have the mish function, you just need the gradient and you are good to go.

reunanen replied:
Where is `CUDNN_ACTIVATION_MISH` even available? Downloaded latest cuDNN, but couldn't find anything.

thebhatman replied:
Yeah it's not available. I need to change this and use the cuda code only.

reunanen replied:
Fixed by adding a standalone CUDA implementation.

davisking commented on dlib/cuda/cuda_dlib.cu:
```diff
@@ -1351,6 +1351,28 @@ namespace dlib
+            const float p = *pp;
+            for (auto i : grid_stride_range(0, n))
+            {
+                auto delta = 2*std::exp(s[i]) + std::exp(2*s[i]) + 2;
```
This is more efficiently represented as:
```
    auto e = std::exp(s[i]);
    auto delta = 2*e + e*e + 2;
```

reunanen replied:
Fixed.

davisking commented on dlib/test/dnn.cpp:
```diff
@@ -1832,6 +1832,12 @@ namespace
+            mish_ l;
+            auto res = test_layer(l);
+            DLIB_TEST_MSG(res, res);
+        }
```
On second thought, please also add a test that `cuda::mish()` matches the outputs of `cpu::mish()`.  For example, like is done for `cuda::softmax_all()` in this file as well.

reunanen replied:
Added a simple test.

reunanen commented:
I wanted to test Mish, so tried to fix the remaining issues in [this PR](https://github.com/thebhatman/dlib/pull/1). (It did take me quite a while to debug the test failure.)

@thebhatman, please consider merging my PR to your branch (I guess this PR will then update automatically).

thebhatman commented:
@reunanen Thanks a lot for your effort and for helping me out.

reunanen commented:
> I just ran the unit tests (via ./dtest --test_dnn) with dlib built without cuda and got this error:

@davisking , this should now work.

reunanen commented:
FYI, I [updated the semantic-segmentation example to use Mish](https://github.com/reunanen/dlib/commit/667c2ad764e360239c39eb87a512a7a9b350ce84) (instead of ReLU). The results are pretty much the same as those reported in #1600 – I wouldn't assume that the small differences are statistically significant:

![image](https://user-images.githubusercontent.com/2297572/72210464-ad109180-34c4-11ea-8fc3-91d849293c3e.png)

```
step#: 119268  learning rate: 0.0001  average loss: 0.0229698    steps without apparent progress: 4026
step#: 119375  learning rate: 0.0001  average loss: 0.0237883    steps without apparent progress: 4130
step#: 119482  learning rate: 0.0001  average loss: 0.0235407    steps without apparent progress: 4234
step#: 119591  learning rate: 0.0001  average loss: 0.0246943    steps without apparent progress: 4364
Saved state to pascal_voc2012_trainer_state_file.dat
step#: 119693  learning rate: 0.0001  average loss: 0.0214533    steps without apparent progress: 1016
step#: 119796  learning rate: 0.0001  average loss: 0.0256509    steps without apparent progress: 4572
step#: 119905  learning rate: 0.0001  average loss: 0.0234571    steps without apparent progress: 4676
step#: 120014  learning rate: 0.0001  average loss: 0.0250673    steps without apparent progress: 4780
step#: 120123  learning rate: 0.0001  average loss: 0.0235702    steps without apparent progress: 4884
step#: 120232  learning rate: 0.0001  average loss: 0.0233183    steps without apparent progress: 4988
Saved state to pascal_voc2012_trainer_state_file.dat_
saving network
Testing the network...
train accuracy  :  0.991318
val accuracy    :  0.847972
```

davisking commented:
Well, I merged this pull request.  But did it in a way github doesn't like apparently.  It's merged though ¯\_(ツ)_/¯

davisking commented:
It sort of bothers me that github is saying this isn't merged.  I don't want to close it and make it look like it wasn't merged.  You want to try rebasing and maybe I can get github to mark it as merged?

thebhatman commented:
I have resolved the merge conflicts. Maybe you can merge the PR now. Or maybe a rebase on master by dropping the previous merge commits and merging this might help.

davisking commented:
Ha, it’s fine. Merged it :)
